H2/scheduler

Klassikcat opened the pull request:
@simnalamburt

This pull request introduces a lottery scheduling mechanism to the `xv6` kernel, along with updates to project guidelines. The most important changes include the addition of lottery scheduling logic in the `scheduler` function, the introduction of a random number generator for ticket selection, and updates to the `proc` and `cpu` structures to support this functionality. Additionally, a `.copilot-instructions` file was added to provide specific usage guidelines for Copilot in this project.

### Lottery Scheduling Implementation:

* **Lottery scheduling logic in `scheduler`:** The `scheduler` function was modified to implement lottery scheduling. Processes are now selected based on a random lottery ticket system, ensuring fairness in CPU allocation. (`kernel/proc.c`, [kernel/proc.cR438-R519](diffhunk://#diff-f06ba4e6ae5b13cb787bad05af2c231bcf826410f25bfd78b18774115f0d20a8R438-R519))
* **Random number generator (`tick_random`):** A new function, `tick_random`, generates random numbers for selecting the winning lottery ticket. This function uses a per-CPU random seed for generating pseudo-random numbers. (`kernel/proc.c`, [kernel/proc.cR438-R519](diffhunk://#diff-f06ba4e6ae5b13cb787bad05af2c231bcf826410f25bfd78b18774115f0d20a8R438-R519))

### Data Structure Updates:

* **Added `lottery_ticket_num` to `struct proc`:** Each process now has a `lottery_ticket_num` field to store the number of tickets it holds for the lottery scheduler. (`kernel/proc.h`, [kernel/proc.hR95](diffhunk://#diff-3e06d244cbfbd7a5db4cc7dabc8a9b07e38ef4a345687c6bce70ef4f40268a38R95))
* **Added `rand_seed` to `struct cpu`:** A new field, `rand_seed`, was added to the `cpu` structure to maintain a per-CPU random seed for the random number generator. (`kernel/proc.h`, [kernel/proc.hR27](diffhunk://#diff-3e06d244cbfbd7a5db4cc7dabc8a9b07e38ef4a345687c6bce70ef4f40268a38R27))

### Project Guidelines:

* **Added `.copilot-instructions` file:** This file provides detailed guidelines for using Copilot in the project, specifying when and how its suggestions should be used, with a focus on low-level systems programming and adherence to `xv6` coding conventions. (`.copilot-instructions`, [.copilot-instructionsR1-R18](diffhunk://#diff-dddce23dfe7f4cc3aa8136ea4bfe6f5aa866c26e64cb9cfd04f9dc62a301d582R1-R18))

Klassikcat committed:
[REVERT] revert to global lock on logic
[INPROGRESS] implementing lottery tickets

[ADD] add lottery scheduling with optimistic-retry

[ADD] add lottery scheduling

[FIX] fix comiler warning

[FIX] remove unnecessary else condition

[FIX] process ticket assginment location: sched_lottery_ticket >
allocproc

[FIX] fix compile error

[FIX] fix concurrency issue

[FIX] fix minor concurrency issues

[ADD] add per-cpu random seed

[FIX] move intr to end of loop

[ADD] add lazy seed init

[FIX] fix wrong compare

[FIX] fix wrong cumsum

[FIX] wrong cumsum: modular: total_ticket_num - 1

[REVERT] revert to global lock on logic itself

[FIX] fix wrong lock order on w's state is not RUNNABLE

[ADD] add ascii art on main.c

Klassikcat commented on kernel/proc.c:
```diff
@@ -434,51 +435,88 @@ wait(uint64 addr)
+    w = 0;
+    for(int i = 0; i < total_running_procs; i++) {
+      cumsum += runnable_procs[i]->lottery_ticket_num;
+      if (winner_ticket <= cumsum) {
```
Resolved [7b79816](https://github.com/Klassikcat/xv6-riscv/pull/3/commits/7b7981651fc88fecd0e97e022411023d7361abeb)

Klassikcat commented on kernel/proc.c:
```diff
@@ -434,51 +437,91 @@ wait(uint64 addr)
 {
   struct proc *p;
   struct cpu *c = mycpu();
+  struct proc *runnable_procs[NPROC];
+  int total_running_procs;
```
TODO: set runnable procs global variables

Klassikcat commented:
TODO:
1. Add system call `set_tickets`
1. Add test code

simnalamburt commented on kernel/proc.c:
```diff
@@ -25,6 +25,7 @@ extern char trampoline[]; // trampoline.S
 // memory model when using p->parent.
 // must be acquired before any p->lock.
 struct spinlock wait_lock;
+struct spinlock scheduler_global_lock;
```
이 락은 어느 자원을 보호하기위해 존재하나요? 따로 보호할 자원이 없다면 락을 걸지 말고 여러 CPU가 동시에 스케줄러를 실행할 수 있도록 허락하는게 나아보입니다.

Klassikcat replied:
생각해보니 따로 보호할 자원이 없기도 하고, 어차피 뒤에서 상태를 다시 체크하니 global lock으로 cpu 하나에서 실행하도록 제한하는 것보단 여러 cpu에서 동시에 스케쥴러를 실행하도록 하는 게 나을 것 같네요. 감사합니다. 

추가로 궁금증이 하나 있습니다.

현재 제가 짠 로직에서는 당첨자를 뽑은 뒤에는 락을 건 뒤 상태를 재확인하는 단계를 거치는데, 만약 프로세스의 상태가 바뀌어서 RUNNABLE이 아니게 되면(SLEEP이라던가...) retry를 하게 되고, 그렇게 되면 cpu 낭비가 심하지 않을까 고민됩니다. 이 문제로 고민을 해봤는데, 혹시 일반적인 운영체제에서는 이런 비효율을 어떻게 해결하나요?

simnalamburt commented on kernel/proc.h:
```diff
@@ -91,6 +92,7 @@ struct proc {
   int killed;                  // If non-zero, have been killed
   int xstate;                  // Exit status to be returned to parent's wait
   int pid;                     // Process ID
+  int lottery_ticket_num;      // Number of lottery tickets for scheduling
```
현재 구현상으로는 항상 10으로 고정된걸로 보입니다. 10으로 고정된다면 없애는게 나아보이고, 10으로 고정을 시키지 않게된다면 그때 추가하는게 나아보여요

Klassikcat replied:
수정했습니다!

simnalamburt commented on kernel/proc.h:
```diff
@@ -24,6 +24,7 @@ struct cpu {
   struct context context;     // swtch() here to enter scheduler().
   int noff;                   // Depth of push_off() nesting.
   int intena;                 // Were interrupts enabled before push_off()?
+  unsigned int rand_seed;     // Random seed for lottery scheduling.
```
시드를 전역변수로 두지않고 Per-CPU로 두어서, CPU들이 서로 간섭하지 않고 병렬적으로 랜덤을 계산할 수 있게 한것 좋습니다 👍 근데 랜덤 시드는 랜덤함수에 맨 처음 넣는 값을 의미하는 단어여서, 이 변수 이름은 `rand_state` 같은걸로 바꾸는게 더 나아보입니다

Klassikcat replied:
수정했습니다! 앞으로 변수 이름은 좀 더 생각하고 작성하도록 하겠습니다!

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +437,91 @@ wait(uint64 addr)
+  int total_ticket_num;
+  int cumsum;
+  int winner_ticket;
+  struct proc *w;
```
변수 선언을 함수 맨 위에 몰아서 하실 필요는 없습니다. 변수선언을 최대한 미루시는게 일반적으로 좋은 practice입니다

Klassikcat replied:
수정했습니다! c에서도 최대한 늦게 변수 선언을 하는 게 좋은 관습이군요.

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +437,91 @@ wait(uint64 addr)
+  }
+  
+  c->rand_seed = c->rand_seed * 1103515245 + 12345;
+  return (unsigned int)(c->rand_seed / 65536) % 32768;
```
지금 수식만 보면 rand_seed가 `* 1103515245 + 12345` 를 거듭하다가 오버플로우로 낮은 확률로 우연히 0이 될 수 있는데, 그때에 임의로 다시 시드 초기화가 발생할 수 있습니다.

함수 하나에서 초기화와 get random을 동시에 하지 말고, init_random() 함수와 get_random() 를 나누는게 좋아보입니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +437,91 @@ wait(uint64 addr)
       asm volatile("wfi");
+      continue;
+    }
+    winner_ticket = tick_random() % total_ticket_num; // 0~(total_ticket_num-1)
```
tick_random() 이 [0, 32768) 범위의 정수여서 이걸 [0, total_ticket_num) 범위로 바꾸시려고 이렇게 하신거같은데, 이러면 uniform하지 않아집니다. 아래 링크를 한번 봐보시고, 어떻게하면 uniform을 유지하면서 범위를 줄일 수 있는지 고쳐보셔요.

<https://github.com/simnalamburt/snippets/blob/c1d19d7b19/python/uniformness-of-random/main.ipynb>

Klassikcat replied:
분포가 uniform하지 않으면 lottery ticket의 가장 중요한 원칙 중 하나인 공정성 측면에서 크게 문제가 되니 해당 부분에 대해 고민을 해봐야겠네요. 링크 참조하면서 어떻게 uniform하게 나올 수 있도록 만들 수 있을지 고민을 해봐야 할 것 같습니다. 감사합니다.

simnalamburt commented:
그리고 https://github.com/remzi-arpacidusseau/ostep-projects/tree/master/scheduling-xv6-lottery 이거 보시면 그래프도 그리게되어있는데 그래프도 첨부 부탁드려요

Klassikcat committed:
[FIX] remove global lock since no global resources to protect

Klassikcat committed:
[FIX] fix name on cpu state in line 27: rand_seed -> rand_state

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +434,87 @@ wait(uint64 addr)
 }
 
+int
+tick_random(void) 
```
tick_random() 함수는 proc.c 에서만 사용되는것같은데, 이런경우 `static tick_random(void)` 로 선언해주시면 좋습니다

Klassikcat committed:
[ADD] add settickets and sys_getpinfo

Klassikcat committed:
[FIX] fix wrong syscall.h

Klassikcat committed:
Merge branch 'riscv' into H2/scheduler

Klassikcat committed:
[ADD] add times_executed on pstat and proc struct

Commit:
[ADD] add notebook: multiplication with right shift note

Commit:
[FIX] fix implementation

Commit:
[ADD] add unbiased random

Klassikcat committed:
[ADD] add test code, [TODO] create getpinfo in user space to observe even process distribution

Klassikcat committed:
[TODO] implement get-pstat

Klassikcat committed:
[ADD] system call getpinfo and user getpinfo

Klassikcat committed:
[TODO] complete test code

Klassikcat committed:
[FIX] disabling formatter [REMOVE] python

Klassikcat committed:
[ADD] lsp settings for vim

Klassikcat committed:
[ADD] add user function getpinfo(for testing)

Klassikcat committed:
[ADD] system call settickets and getpinfo

Klassikcat committed:
[ADD] add cpu tasks and graph

Klassikcat committed:
[FIX] execuatble getpinfo and get_schedlog

Klassikcat committed:
[REMOVE] unnecessary files

Klassikcat committed:
[FIX] fix scheduler bug: cumsum location

Klassikcat committed:
[FIX] add infinite scheduler testing code

Klassikcat committed:
[TODO] implement expected ratio code

Klassikcat committed:
[FIX] remove unused tickets

simnalamburt commented:
원하시는 수정이 끝나셨을때 다시 request review 버튼 눌러주시면 살펴보겠습니다!

Klassikcat committed:
[ADD] scheduler logs

Changelog:
 Add Scheduler logs to monitor ticket number per process and execution
times so that lottery ticket works properly

Klassikcat commented:
@simnalamburt 스케줄러의 정상 작동을 확인했으며, 스케줄러에 버그가 있던 부분을 수정하였습니다.

프로세스 티켓의 누적합을 구하는 `cumsum`의 위치가 루프 안에 있어야 하는데, 루프 밖에 있어 루프마다 초기화되지 않는 이슈가 있었습니다. 특정 루프 이후로는 cumsum이 total_ticket_num보다 더 커지게 되고, `winner_ticket < cumsum`이 언제나 참이 돼서 첫 번째 runnable 프로세스만 당첨되어 나머지 프로세스가 기아가 되는 문제가 발생했습니다. 해당 부분은 `6728049` 커밋에서 수정하였습니다.

스케줄러 로그는 아래와 같습니다.

```

xv6 kernel is booting

██████╗  █████╗ ███╗   ██╗██████╗  ██████╗ ███╗   ███╗ █████╗ ███╗   ██╗
██╔══██╗██╔══██╗████╗  ██║██╔══██╗██╔═══██╗████╗ ████║██╔══██╗████╗  ██║
██████╔╝███████║██╔██╗ ██║██║  ██║██║   ██║██╔████╔██║███████║██╔██╗ ██║
██╔══██╗██╔══██║██║╚██╗██║██║  ██║██║   ██║██║╚██╔╝██║██╔══██║██║╚██╗██║
██████╔╝██║  ██║██║ ╚████║██████╔╝╚██████╔╝██║ ╚═╝ ██║██║  ██║██║ ╚████║
╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝╚═════╝  ╚═════╝ ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝  ╚═══╝
                                B A N D O                          
                                  M A N                          
init: starting sh
$ get_schedlog
Executed fib_infinite pid=4 with n=40 and sleep=300
Executed fib_infinite pid=6 with n=40 and sleep=300
Executed fib_infinite pid=5 with n=40 and sleep=300
Process 4: tickets=30, executions=8
Process 5: tickets=20, executions=6
Process 6: tickets=10, executions=2
Process 4: tickets=30, executions=169
Process 5: tickets=20, executions=102
Process 6: tickets=10, executions=49
Process 4: tickets=30, executions=311
Process 5: tickets=20, executions=226
Process 6: tickets=10, executions=94
Process 4: tickets=30, executions=460
Process 5: tickets=20, executions=324
Process 6: tickets=10, executions=149
Process 4: tickets=30, executions=614
Process 5: tickets=20, executions=435
Process 6: tickets=10, executions=190
Process 4: tickets=30, executions=761
Process 5: tickets=20, executions=543
Process 6: tickets=10, executions=240
Process 4: tickets=30, executions=918
Process 5: tickets=20, executions=634
Process 6: tickets=10, executions=293
Process 4: tickets=30, executions=1064
Process 5: tickets=20, executions=734
Process 6: tickets=10, executions=351
Process 4: tickets=30, executions=1222
Process 5: tickets=20, executions=833
Process 6: tickets=10, executions=394
Process 4: tickets=30, executions=1374
Process 5: tickets=20, executions=949
Process 6: tickets=10, executions=432
```

simnalamburt commented on kernel/pstat.h:
```diff
@@ -0,0 +1,10 @@
+#define NPROC 64 
+#endif
+
+struct pstat {
+  int inuse[NPROC];   // whether this slot of the process table is in use (1 or 0)
+  int lottery_tickets[NPROC]; // the number of tickets this process has
+  int pid[NPROC];     // the PID of each process 
+  int times_executed[NPROC];
+};
```
inuse는 boolean같은데 4byte int 말고 _Bool 써서 용량 아끼시죠

저 이거 오래돼서 잘 기억이 안나는데, struct pstat의 모양을 괴제에서 누군가가 강제하던가요? struct pstat 안에 int[N]을 네개 넣는게 아니라 struct pstat 을 int를 네개 넣고 배열 struct pstat[N]을 다루는쪽이 Taste도 더 낫고, 캐시프렌들리할것같은데, 꼭 이렇게 생겨야했던건지 아닌지 기억이 안나서요.

Klassikcat replied:
아마 pstat struct에 대해서 OSTEP 책에서 가이드가 있었던 것으로 기억합니다. struct pstat에는 어떤 어떤 attribute가 있어야 하고 하는 것까진 기억나는데, 책 확인해보고 회신할게요

Klassikcat replied:
ostep.org는 지금 제 인터넷이 문제인지 접근이 안돼서 확인하진 못했는데, 다른 곳에서 같은 lab 과제를 보면 그걸 강제하진 않는 것 같습니다. 아마 가이드라인 정도가 아닌가 싶은데, 해당 부분도 반영해서 작업해보겠습니다. [참조](https://intra.ece.ucr.edu/~cong/teaching/UCR/AOS/labs/lab2.pdf)

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +494,96 @@ wait(uint64 addr)
+  struct proc *runnable_procs[NPROC];
+  int total_ticket_num;
+  int cumsum;
+  int winner_ticket;
```
(중복리뷰) 변수선언은 최대한 늦게, 변수 사용하기 직전에 해주세요

Klassikcat replied:
어 이 부분 수정을 안했었네요??? 다음 리뷰때 반영하겠습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +494,96 @@ wait(uint64 addr)
 
   c->proc = 0;
+  init_random();
+  struct proc *w;
```
루프 바깥에서 선언한 이유가 무엇인가요?

Klassikcat replied:
저도 작업한 지 오래돼서 확인해 봤는데, cpu 구조체에서 rand_state가 없을 때에만 init_random을 돌리기 때문에 그렇습니다. 그리고 현재 cpuid와 ticks를 활용하여 rand_state를 만드는지라 루프마다 새로 만들면 난수 품질이 떨어질 것 같아서 그때 루프 밖에다가 넣은 것 같습니다.

다만 확인했을 때 생각보다 초기화 함수의 품질이 낮은 것을 확인해서, 이 부분은 다음 리뷰 때 반영하여 수정하겠습니다.

simnalamburt replied:
수정안되어있음

Klassikcat replied:
init_random 말씀하시는 거 맞죠? 그거 아마 작업해두고 다른 PC에서 push 안한 것 같은데, 별도 PR에다가 넣어서 올릴게요 아마 그거 코드 중복 이슈도 있을거라서

simnalamburt commented on kernel/syscall.h:
```diff
@@ -20,4 +20,6 @@
\ No newline at end of file
+#define SYS_getreadcount 22
+#define SYS_getpinfo 23
+#define SYS_settickets 24
```
LF at EOF 추가

Klassikcat replied:
으악 린터에 중독되는 바람에 그만 

확인해보니 LF를 추가해야 하는 이유가 POSIX 표준이라서 그런 거였네요. 해당부분 커밋 반영하도록 하겠습니다

simnalamburt replied:
안되어있어요

Klassikcat replied:
포매터가 지워버렸네요; 새 RR 올려서 수정하겠습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +494,96 @@ wait(uint64 addr)
       asm volatile("wfi");
+      continue;
+    }
+    winner_ticket = ((uint64)get_random() * total_ticket_num) >> 15; // output of get_random() is 15-bit int
```
이렇게 걍 나누면 total_ticket_num 이 2**15 의 약수일때에만 유니폼합니다.

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +494,96 @@ wait(uint64 addr)
       intr_on();
       asm volatile("wfi");
+      continue;
+    }
+    winner_ticket = ((uint64)get_random() * total_ticket_num) >> 15; // output of get_random() is 15-bit int
+    w = 0;
+    for(int i = 0; i < total_running_procs; i++) {
+      cumsum += runnable_procs[i]->lottery_tickets;
+      if (winner_ticket < cumsum) {
+        w = runnable_procs[i];
+        break;
+      }
+    }
+    acquire(&w->lock);
+    if (w->state != RUNNABLE) {
```
w에 대한 null check를 안한 이유가 논리적으로 w가 무조건 선정될거라고 생각하셔서 그런거같은데, 지금 이 루프에선 w가 락이 걸려있지 않기때문에 w가 선정되지 않을 수 있습니다.

앞 루프에서 runnable_procs 를 모을때, 락을 건채로 모아야합니다

simnalamburt reviewed (changes requested):
제가 한 리뷰는 GPT-5.3-Codex xhigh 도 아마 찾아줄거같습니다. 선단모델로(선단이어야함) 리뷰를 받아보세요.

Klassikcat committed:
[FIX] scheduling logic change

from: LCG + 15bit reduction
to:  Weyl sequence + lowbias32 finalizer

Klassikcat commented:
@simnalamburt 코드 리뷰 결과를 종합하여 아래와 같이 변경하였습니다. 선단모델(GPT 5.5, Claude Opus 4.8)을 통해 추가적으로 리뷰를 받았습니다.

## 스케쥴러 bias 관련

스케쥴러에서 bias가 있을 수 있다는 것을 확인하여, PRNG를 15-bit source에서 32-bit source으로 올리고, `>> 32` multiply-high 연산을 통해 상대적인 편향을 줄였습니다. 

추가로, LCG 형태의 PRNG에서 Weyl 순열과 lowbias32 mixer를 사용하였습니다. state를 특정 상수를 통해 32-bit 숫자 공간에서 전진시키고, bit shift 연산을 한 뒤, XOR 연산을 통해 비트를 바꾼 뒤 결과를 저장합니다. 이를 반복하는 형태로 의사 랜덤 숫자를 출력합니다.

기존 LCG는 input과 output의 관계가 선형적이기 때문에 출력한 랜덤 숫자 하나와 바로 다음에 출력하는 랜덤 숫자간의 관계가 선형적이고 명확하여 랜덤함이 떨어지는 반면, Weyl + lowbias32는 bit shift + XOR이 input과 output의 관계를 비선형적으로 만들어 조금 더 랜덤해집니다.

## w에 대한 null check 부분

runnable procs를 모을 때 락을 건 채로 모을지, 아니면 w가 선정되지 않을 수 있을 것을 가정하고 null check를 할지 고민이 조금 있었는데, 결론적으로 w에 대한 null check가 현재 코드 구조에선 더 빠른 해결책이 될 것이라 생각하여 최종적으로 w에 대해서 null check를 하는 식으로 구현하였습니다. 

## 기타 수정사항

_Bool로 했을 때 lsp가 에러를 뿜어서 int로 한 거였는데, false alarm인 것으로 확인되어 _Bool로 변경했습니다.

simnalamburt commented:
?

Klassikcat reviewed (commented):
댓글 달았습니다

simnalamburt commented on kernel/proc.c:
```diff
@@ -20,15 +21,72 @@ static void freeproc(struct proc *p);
+uint64
+sys_getpinfo(void) {
+  struct pstat pstat;
+  struct proc *p;
+  uint64 addr;
+
+  argaddr(0, &addr);
+  if (addr == 0) {
+    return -1;
+  }
+
+  int i = 0;
+  for(p = proc; p < &proc[NPROC]; p++) { // process table iter
```
```suggestion
  uint64 addr;

  argaddr(0, &addr);
  if (addr == 0) {
    return -1;
  }

  int i = 0;
  for(struct proc *p = proc; p < &proc[NPROC]; p++) { // process table iter
```

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +495,109 @@ wait(uint64 addr)
+get_random(void)
+{
+  struct cpu *c = mycpu();
+  uint x;
+
+  c->rand_state += 0x9e3779b9U;
+  x = c->rand_state;
```
```suggestion
  c->rand_state += 0x9e3779b9U;
  uint x = c->rand_state;
```

simnalamburt commented on kernel/proc.c:
```diff
@@ -434,51 +495,109 @@ wait(uint64 addr)
-    for(p = proc; p < &proc[NPROC]; p++) {
+    intr_on(); // Enable interrupts to allow preemption
+    int total_running_procs = 0;
+    uint total_ticket_num;
+    total_ticket_num = 0;
+    uint cumsum;
+    cumsum = 0;
```
<img height="200" alt="Image" src="https://github.com/user-attachments/assets/c6c3fcf8-d857-452b-90be-9f1a51fcc5b9" />


```suggestion
    uint total_ticket_num = 0;
    uint cumsum = 0;
```

Klassikcat replied:
C 문법이 익숙하지 않아서 xv6 내의 다른 코드의 용법을 참조해서 썼는데, 그런 걸 보면서 쓰다보니 이런 코드가 꽤 나왔습니다. 예를 들어 int wait(uint64 addr) 같은 코드를 보면 

```c
int
wait(uint64 addr)
{
  struct proc *pp;
  int havekids, pid;
  struct proc *p = myproc();

  acquire(&wait_lock);

  for(;;){
    // Scan through table looking for exited children.
    havekids = 0;
    for(pp = proc; pp < &proc[NPROC]; pp++){
      if(pp->parent == p){
        // make sure the child isn't still in exit() or swtch().
        acquire(&pp->lock);
...
```

이런 식으로 코드가 짜여져있어서 그걸 보고 스타일을 맞췄었습니다

말씀하신대로 코드를 짜는 게 좋은데 왜 이런 식으로 된 코드와 for문 내부에서 변수를 선언하는 코드가 섞여있나 싶어서 찾아봤는데, 이게 c89에서의 제약이라고 하더라고요. c89에선 블록의 시작 지점에서만 변수를 선언할 수 있었다고 합니다.

```c
int
func(void) {
 int var = 0;
 int val;
 // exec
 { 
    struct something;
...
```

이런 식으로요. 이게 c99 와서 완화가 됐다고 합니다. 이런 제약이 있어서 저런 식으로 했던 걸 모르고 그냥 대충 따라서 했으니, 자연스럽게 bad code가 나왔던 거고요... 

찾아서 별도의 브랜치로 올려서 따로 PR 올리겠습니다

simnalamburt reviewed (commented):
그냥
```
int a = 0;
```
이러면 되는걸 굳이 굳이
```
int a;
a = 0;
```
이렇게 하는 코드가 엄청 많고 변수 선언을 전반적으로 다 사용하는 시점이 아니라 한참 전에 미리 해두시는데 그렇게 하시는 이유가 궁금합니다.
